### PR TITLE
fix: follow the query-pattern operators a metadata callback runs

### LINKED ISSUE #2287: BESG004 does not follow five more implicitly-invoked members
## Summary

BESG004 walks what a metadata or execution callback runs *without naming it* — a constructor, a user-defined operator or conversion, an indexer, and (since #2221) the `Add` a collection initializer runs. Five more shapes in that same category are still not followed, so a callback can reach mutable static state through any of them and the rule stays silent while the plan key stays the same.

None of these is a new regression; they are gaps the rule has always had. The reason to close them is the one that motivated the collection-initializer fix: the shape immediately beside them **is** followed, so the followed shape is evadable by rewriting it. `new Builder(value)` was walked while `new Builder { value }` was not — two characters. The five below have the same property.

## The shapes

Each was built as a compiling metadata callback and run through the analyzer; all five are silent.

**1. Object-initializer property setter** — `new Holder { Source = value }` where the setter reads a mutable static.
`GetReceiver` (`MetadataCallbackPurityAnalyzer.cs:1294-1301`) derives a receiver only from a `MemberAccessExpressionSyntax`, `MemberBindingExpressionSyntax` or element access. A bare `Source` whose parent is the initializer's assignment yields no receiver, so `FollowPropertyAccess` bails at `FollowReceiverCreation`.

**2. Indexer assignment inside an object initializer** — `new Table { [0] = value.X }`.
`ImplicitElementAccessSyntax` is not `ElementAccessExpressionSyntax`, so the existing indexer arm never sees it.

**3. Collection expression** — `BoundsBuilder b = [x, y];`
This one needs a **different** approach from the collection-initializer fix: `GetCollectionInitializerSymbolInfo` does not answer for it. The operation is an `ICollectionExpressionOperation` whose `ConstructMethod` is the parameterless constructor, with the `Add` calls appearing only in the lowered form — so it needs `GetOperation` and the element operations.

**4. `with` expression** — a record whose user-written copy constructor reads a mutable static.
`GetSymbolInfo` on the `with` yields nothing, and the operation's `CloneMethod` is the compiler-generated `<Clone>$`, so the user's copy constructor is one hop further.

**5. `foreach` over a user `GetEnumerator`/`MoveNext`/`Current`, and deconstruction via a user `Deconstruct`.**
Both need dedicated APIs the walk does not call: `GetForEachStatementInfo` and `GetDeconstructionInfo`.

**6. Query expression** — `from x in new BoundsQuery(r) where true select x`.
The compiler invokes the query type's source-visible `Where` and `Select` without spelling their names, and
the unnamed-invocation filter ignores query clauses. Roslyn selects the operators, so they are resolvable —
`GetSymbolInfo` on the clause node returns the chosen method — but nothing asks. Reported by the PR reviewer
on #2221 after this issue was filed.

## Why it matters

Structural identity is the callback's `MethodInfo` (`OpaqueRenderDescription.cs:1931`), so two recordings that answer differently still share a compiled plan. That is the whole reason BESG004 exists. For the forward-bounds direction the drift is not silent — `RegionAnalyzer.cs:566-577` re-derives and throws, and it is not Debug-gated — but the hit-test contract and every backward mapping (`GetRequiredInputBounds`, `RenderInputDemandContract`) have no runtime check in any configuration (`SymbolicMetadataCrossCheck.cs:31-46`).

## Realistic severity

Low individually. The contract callbacks are `Func<Rect, Rect>`, `Func<EffectiveScale, EffectiveScale>`, `Func<RenderScaleContext, float>` and `Func<RenderHitTestContext, Point, bool>`, whose natural bodies are arithmetic on the argument; a builder is contrived there. The statement-bodied execution callbacks on `PaintedSourceDefinition` / `TargetCommandDefinition` / `GeometryDefinition` make shapes 1, 2 and 3 more plausible. Note also that a `new List<T> { … }` is harmless because `List<T>.Add` has no source in the compilation and the walk stops there without reporting — the gap only bites for a user-declared type in the same compilation.

## Suggested approach

Shapes 1, 2, 4 and 5 are each a small addition to `FollowUnnamedInvocation` in the same style as the existing arms, and could land together. Shape 3 is separate work because it needs the operation API rather than the syntax/symbol queries the rest of the walk uses — worth deciding whether to introduce that dependency at all before writing it.

Each addition needs a positive test **and** a negative control showing the arm follows the member's body rather than reporting the shape's existence, plus an in-tree sweep (`--no-incremental` rebuild of `Beutl.Engine`, `Beutl.ProjectSystem`, `Beutl.NodeGraph`) confirming nothing newly reports. Note a probe file named `*.g.cs` is skipped by `ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None)`, so a sweep probe must avoid that suffix or it proves nothing.

Any of these that newly reports is a widening of BESG004's break and needs the `BREAKING CHANGE:` note updated accordingly.

---

## Independently reported

Shape 5 was raised again by the PR reviewer on #2221 after this issue was filed, with the same diagnosis and the same remedy:

> When a static metadata callback iterates a source-defined enumerable, `foreach` implicitly invokes `GetEnumerator`, `MoveNext`, `Current`, and `Dispose`, but this unnamed-invocation filter ignores `ForEachStatementSyntax`. If one of those members reads mutable static state — for example, `MoveNext` uses a static threshold — BESG004 remains silent and the unchanged callback identity can reuse stale bounds or hit-test metadata. Resolve the protocol members via Roslyn's foreach statement info and follow their source bodies.

It adds one detail worth carrying into the fix: the protocol is four members, not three — `GetEnumerator`, `MoveNext`, `Current` **and `Dispose`** — and `GetForEachStatementInfo` returns all of them, so the arm should follow each rather than only the obvious two.

Two independent reports of the same gap is a reasonable signal that `foreach` is the most reachable of the five, and a case for doing it first.

### PR DESCRIPTION
**Pull request by yuto-trd**
## Description

Closes the last unfollowed shape in #2287: a **query expression**.

A query spells none of the methods it runs — every clause is rewritten into a call the binder picks off the source's own type by the query pattern — so `FollowUnnamedInvocation` never saw them. A metadata callback could reach mutable static state through a source-declared query operator and BESG004 stayed silent while the plan key stayed the same, which is exactly the hole the rule exists to close.

- `QueryClauseSyntax` is resolved through `GetQueryClauseInfo`, following **both** of its answers: an explicitly typed `from` runs a `Cast` ahead of the operator the clause itself is.
- `SelectOrGroupClauseSyntax` and `OrderingSyntax` are resolved through `GetSymbolInfo`, which names the `Select`/`GroupBy`/`OrderBy`/`ThenBy` the clause does not.
- Operators are taken as the binder already chose them, so an instance operator and an extension one both arrive resolved, and a generic one arrives with its arguments inferred.
- A degenerate `select` — one handing back the range variable of a query that has any other clause — is *removed* by the compiler rather than rewritten, answers with nothing, and is followed nowhere. A LINQ-to-objects query answers with `Enumerable`'s methods, which have no source here and stop the walk as any other such callee does.

### Virtual dispatch on the query source (review follow-up, `6183defb9`)

The first revision followed the operator the binder bound against the source's **static** type, which a review caught. Measured both directions with a base-declared local holding a derived instance:

| shape | first revision | now |
|---|---|---|
| impure base, pure override | reported — **false positive** | not reported |
| pure base, impure override | not reported — **miss** | reported |
| receiver whose making is unreadable | reported as bound | unchanged |

An instance call named outright already resolves this through `FollowReceiverCreation`/`RunsAsMade`; the query arm now takes the same path. Only the **first** operator runs on a receiver the source names — the `from` expression — and which clause carries it depends on the clauses before it, so the whole query is read at once in rewrite order (document order, which for a query is the same thing) rather than clause by clause. Later operators run on values no creation here makes and are followed as bound. A query written inside another is left for the walk to reach as the expression it is, so its first operator resolves against its own source.

### A typed inner clause's `Cast` (review follow-up, `e62ddfa25`)

A second `from` or a `join` runs its `Cast` on the sequence **that clause** introduces, not on the one the chain has reached. The first pass handed every `Cast` the query's own `from` expression, which broke two things at once, both reproduced before the fix:

- the `Cast` was resolved against the wrong receiver, and
- the `SelectMany` / `Join` after it — the first operator that really does run on the query's source — lost its resolution and was followed as bound, missing an impure override on the source's concrete type.

`GetQueryOperators` now hands each `Cast` back with the source its own clause names (`FromClauseSyntax.Expression`, `JoinClauseSyntax.InExpression`), and the chain's first receiver is consumed only by a `Cast` on the query's own `from` expression.

`FollowReceiverCreation` / `GetReceiverCreation` were split so a bare receiver expression can be used; the named-call path is unchanged and its tests are untouched.

Where the source's making cannot be read, the operator is still followed as the binder bound it — a virtual override or an interface implementation can replace that body. That is deliberate: declining there would silence every query over a field or a parameter, and would make this the only unnamed-invocation arm that gates, since `FollowIteration`, `FollowDisposal`, `FollowWithExpression` and `FollowDeconstruction` all follow the bound declaration too. The limit is pinned by a test and named in the release note (`f514d63a3`); making the whole family dispatch-aware is filed with measurements as #2420.

The other five shapes listed in #2287 (object-initializer property setter, implicit element access, collection expression, `with`, `foreach`/`Deconstruct`) already landed in #2221, so this completes the issue.

## Affected areas

- [x] `Beutl.Engine` (rendering / scene / track) — `Beutl.Engine.SourceGenerators` analyzer
- [ ] `Beutl.ProjectSystem` (project / document persistence)
- [ ] UI (`Beutl.Editor`, `Beutl.Editor.Components`, `Beutl.Controls`)
- [ ] `Beutl.Extensibility` (plugin abstractions)
- [ ] `Beutl.NodeGraph` (node editor)
- [ ] `Beutl.FFmpegIpc` / `Beutl.FFmpegWorker` (media IPC boundary)
- [ ] `Beutl.Api` (server API client)
- [ ] Build / CI / docs only

## Breaking changes

None requiring migration. BESG004 reaches further, which is a widening of what it reports, but:

- nothing in the tree newly reports (sweep below);
- BESG004 is still listed in `AnalyzerReleases.Unshipped.md` with an empty `Shipped.md`, so no released build carries the narrower rule to migrate from.

`AnalyzerReleases.Unshipped.md` is updated to describe the query-pattern operators the walk now follows.

## Test plan

`tests/SourceGeneratorTest/MetadataCallbackPurityAnalyzerTests.cs` — 19 new cases behind an `AnalyzeQuery` helper, matching the existing `AnalyzeIteration` shape: exactly one operator is impure per reported case, and the accepted cases keep the same shape so silence means the walk looked.

Positive (each one **fails on `main` and passes here**):

| case | resolved through |
|---|---|
| `…QueryWhere…` | `GetQueryClauseInfo(...).OperationInfo` |
| `…QuerySelect…` | `GetSymbolInfo(SelectOrGroupClauseSyntax)` |
| `…QueryOrderBy…` / `…QueryThenBy…` | `GetSymbolInfo(OrderingSyntax)` |
| `…QueryCast…` | `GetQueryClauseInfo(...).CastInfo` |
| `…QueryGroupBy…` | `GetSymbolInfo(SelectOrGroupClauseSyntax)` |
| `…QuerySelectMany…` | second `from` |
| `…QueryJoin…` / `…QueryGroupJoin…` | `join` / `join … into` |
| `…QueryExtensionWhere…` | a reduced extension operator, reported as the static method it is |
| `…QueryOperatorAnOverrideReplaces…` | the override that runs, not the declaration the query bound to |
| `…NestedQueryOperatorAnOverrideReplaces…` | an inner query resolving against its own source |
| `…QuerySelectManyAnOverrideReplaces…` | a typed inner `from` not displacing the outer `SelectMany` |
| `…QueryCastAnOverrideReplaces…` | a typed inner `from`'s `Cast` resolving against its own source |

Negative controls (pass both with and without the fix, so the arm answers on what the operator's body does rather than on the clause having been written):

- `AStaticLambdaQueryingAPureSource_IsNotReported`
- `AStaticLambdaWhoseDegenerateQuerySelectIsRemoved_IsNotReported` — the source's `Select` reads a mutable static and is still not reported, because the compiler removed that call
- `AStaticLambdaQueryingAFrameworkSequence_IsNotReported`
- `AStaticLambdaWhoseQueryOperatorAnOverrideReplacesWithAPureOne_IsNotReported` — the base body reads a mutable static and is still not reported, because the override is what runs
- `AStaticLambdaWhoseQueryOperatorRunsOnAReceiverItDidNotMake_IsReportedAsBound` — pins that an unreadable making leaves the operator as the binder bound it

Runs:

- `dotnet test tests/SourceGeneratorTest` — **347 passed, 0 failed**
- `dotnet format --verify-no-changes` on both touched projects — clean

In-tree sweep, `--no-incremental` rebuild of every project that references the analyzer (`Beutl.Engine`, `Beutl.ProjectSystem`, `Beutl.NodeGraph`, plus `samples/PackageSample` and the four test assemblies): **no BESG00x diagnostic anywhere**, all builds succeed.

The sweep is a meaningful negative, not a silent one: a temporary probe (`src/Beutl.Engine/SweepProbe.cs`, deliberately **not** `*.g.cs`, since `ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None)` skips that suffix) holding a `RenderBoundsContract.Create` callback with a query whose `Where` reads a mutable static produced `warning BESG004` on the real build. The probe was removed afterwards and is not part of this PR.

## Fixed issues / References

- Closes #2287

### Out of scope, noted rather than fixed

- The `AnalyzerReleases.Unshipped.md` BESG004 note does not mention the **interpolated string handler** arm either; that arm landed in #2221 and its omission is a separate pre-existing documentation gap, left alone here.
- #2420 tracks dispatch across the whole unnamed-invocation family, raised by review here and fixed for the query arm only.
- #2293 tracks the rest of the implicit-invocation class (list patterns, positional patterns, rebindable `ref` aliases, base-type mutators) and the question of whether the walk should move to `IOperation`. This PR does not answer that; it only closes the shape #2287 names.

### CODE REVIEW

File: src/Beutl.Engine.SourceGenerators/Analyzers/MetadataCallbackPurityAnalyzer.cs [hunk 1]
```diff
@@ -1093,6 +1119,37 @@ private static void FollowInterpolatedStringHandler(
...
+        Dictionary<ISymbol, int> walked,
+        Action<SyntaxNode, string, ISymbol, string> report)
+    {
+        if (chosen is not IMethodSymbol rewritten)
+            return;
+
+        FollowCall(
+            context,
+            rewritten,
+            node,
```
[the remark this answers is not in the document]
Comment (yuto-trd) [reply src/Beutl.Engine.SourceGenerators/Analyzers/MetadataCallbackPurityAnalyzer.cs]: Confirmed and fixed in 6183defb9. Measured both directions against the probe you describe: an impure base with a pure override reported (false positive), and an impure override was missed. The whole query is now read at once in rewrite order, and its first operator — the only one whose receiver the source names — is re-resolved through `GetCreationHeldBy`/`RunsAsMade`, the same path an instance call named outright takes; later operators run on values no creation here makes and stay as bound. Three regression tests cover the override, the pure-override and the unreadable-receiver cases, plus one for a nested query resolving against its own source.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/Beutl.Engine.SourceGenerators/Analyzers/MetadataCallbackPurityAnalyzer.cs [hunk 2]
```diff
@@ -1093,6 +1119,37 @@ private static void FollowInterpolatedStringHandler(
...
+            RunsAStaticMethod(rewritten) ? "static method" : "method",
```
[the remark this answers is not in the document]
Comment (yuto-trd) [reply src/Beutl.Engine.SourceGenerators/Analyzers/MetadataCallbackPurityAnalyzer.cs]: Half fixed, half deliberate. The known-receiver case landed in 6183defb9: the query is now read at once in rewrite order, and its first operator — the only one whose receiver the source names — goes through `GetCreationHeldBy`/`RunsAsMade`, the same path a named instance call takes. Measured: impure base + pure override went from reported (false positive) to silent, pure base + impure override from silent to reported.

The unknown-dispatch case is followed as bound on purpose. Declining there would silence every query over a field or parameter, and would make this the only unnamed-invocation arm that gates — `FollowIteration`, `FollowDisposal`, `FollowWithExpression` and `FollowDeconstruction` all follow the bound declaration too, and a `foreach` over the same base-declared local still reports the base body today. Making that whole family dispatch-aware is a contract change to a rule already on `main`, so it is filed as #2420 with the measurements rather than done here. The limit is pinned by `AStaticLambdaWhoseQueryOperatorRunsOnAReceiverItDidNotMake_IsReportedAsBound` and named in `AnalyzerReleases.Unshipped.md` (f514d63a3).

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/Beutl.Engine.SourceGenerators/Analyzers/MetadataCallbackPurityAnalyzer.cs [hunk 3]
```diff
@@ -1093,6 +1104,103 @@ private static void FollowInterpolatedStringHandler(
...
+            FollowCall(
+                context,
+                source ? RunsAsMade(made, rewritten) : rewritten,
+                node,
+                RunsAStaticMethod(rewritten) ? "static method" : "method",
+                depth,
+                walked,
+                report);
+
+            source = false;
```
[the remark this answers is not in the document]
Comment (yuto-trd) [reply src/Beutl.Engine.SourceGenerators/Analyzers/MetadataCallbackPurityAnalyzer.cs]: Confirmed and fixed in e62ddfa25 — both halves reproduced before the fix. `GetQueryOperators` now hands each `Cast` back with the source its own clause names (`FromClauseSyntax.Expression`, `JoinClauseSyntax.InExpression`), and the chain's first receiver is consumed only by a `Cast` on the query's own from expression, so a typed inner clause no longer displaces the `SelectMany`/`Join` that really is the first operator on the outer source.

Two regression tests, each failing on the previous commit: `…QuerySelectManyAnOverrideReplaces…` (the outer `SelectMany` override was missed) and `…QueryCastAnOverrideReplaces…` (the inner `Cast` was resolved against the outer source). 347 analyzer tests pass and the in-tree sweep across all eight analyzer consumers still reports nothing.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/Beutl.Engine.SourceGenerators/Analyzers/MetadataCallbackPurityAnalyzer.cs [hunk 4]
```diff
@@ -1093,6 +1104,103 @@ private static void FollowInterpolatedStringHandler(
...
+                    body,
+                    node,
+                    depth,
+                    walked,
+                    report);
+            }
+
+            FollowCall(
+                context,
+                source ? RunsAsMade(made, rewritten) : rewritten,
```
[the remark this answers is not in the document]
Comment (yuto-trd) [reply src/Beutl.Engine.SourceGenerators/Analyzers/MetadataCallbackPurityAnalyzer.cs]: Measured, and not changed here — this needs a capability the analyzer has nowhere rather than a fix to a divergence.

Built your exact shape (`Widths items = new OffsetWidths()`, a virtual `Where` returning `this`, an impure `OffsetWidths.Select` override) and ran it in both spellings. The query form reports nothing, and the hand-written `items.Where(...).Select(...)` reports nothing either: `GetReceiverCreation` gets an `InvocationExpressionSyntax` as the receiver, which is neither a creation nor a name holding one, so the named path declines to follow at all. The query arm is already the more generous of the two — it follows the bound declaration where the named path follows nothing.

Propagating the receiver would mean proving an operator returns its own receiver on every path, which is inter-procedural must-alias analysis and is impossible for an operator with no source here (`Enumerable.Where` returns a new iterator, not `this`). That is the same axis as the dispatch question the other two threads raised, so it is recorded on #2420 with this measurement rather than done in this PR.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
